crankshaft/compiler.md

```diff
@@ -0,0 +1,281 @@
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**  *generated with [DocToc](http://doctoc.herokuapp.com/)*
+
+- [v8 Compiler](#v8-compiler)
+  - [Components](#components)
+    - [Base Compiler](#base-compiler)
+    - [Runtime Profiler](#runtime-profiler)
+    - [Optimizing Compiler](#optimizing-compiler)
+    - [Deoptimization Support](#deoptimization-support)
+  - [Optimized Code vs. Inline Caches and Unoptimized Code](#optimized-code-vs-inline-caches-and-unoptimized-code)
+    - [Full Compiler](#full-compiler)
+    - [Inline Caches](#inline-caches)
+    - [Monomorphism vs. Polymorphism](#monomorphism-vs-polymorphism)
+      - [Considerations](#considerations)
+    - [Optimizing Compiler](#optimizing-compiler-1)
+    - [Deoptimization](#deoptimization)
+    - [Causes for Deoptimization](#causes-for-deoptimization)
+      - [Modifying Object Shape](#modifying-object-shape)
+        - [Considerations](#considerations-1)
+  - [Efficiently Representing Values and Tagging](#efficiently-representing-values-and-tagging)
+    - [Considerations](#considerations-2)
+  - [Arrays](#arrays)
+    - [Fast Elements](#fast-elements)
+      - [Characteristics](#characteristics)
+    - [Dictionary Elements](#dictionary-elements)
+      - [Characteristics](#characteristics-1)
+    - [Double Array Unboxing](#double-array-unboxing)
+    - [Typed Arrays](#typed-arrays)
+      - [Float64Array](#float64array)
+    - [Considerations](#considerations-3)
+  - [Resources](#resources)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+# v8 Compiler
+
+## Components
+
+### Base Compiler 
+
+- is used for all code initially
+- generates code quickly without heavy optimizations
+- compilation with the base compiler is very fast generates little code
+
+### Runtime Profiler 
+
+- monitors the running system and identifies hot code
+
+### Optimizing Compiler 
+
+- recompiles and optimizes hot code identified by the runtime profiler
+- uses static single assignment form to perform optimizations
+  - loop-invariant code motion
+  - linear-scan register allocation
+  - inlining.
+- optimization decisions are based on type information collected while running the code produced by the base compiler
+
+### Deoptimization Support 
+
+- allows the optimizing compiler to be optimistic in the assumptions it makes when generating code
+- deoptimization support allows to bail out to the code generated by the base compiler if the assumptions in the
+  optimized code turn out to be too optimistic
+
+## Optimized Code vs. Inline Caches and Unoptimized Code
+
+[watch](http://youtu.be/VhpdsjBUS3g?t=18m53s)
+[watch](http://youtu.be/UJPdhx5zTaw?t=26m30s)
```

Comment:
Can haz more descriptive link text?

Reply:
Comment is on old content, but I carried the way of doing things into the new one as well, so if it's a problem I'd need to fix it there.

The idea is that I have resources with descriptive title at the end of sections or documents.
The little `[watch]` and `[read]` links are always in subsections and targeting exactly the topic discussed, as in if you need more info about what you're just reading, click on it.
They are meant to be small in order not to break the reading flow.

I'm open to suggestions though on how to improve on this.

Reply:
Would the combination of a short link with italicized description work, i.e. like [here](https://github.com/thlorenz/v8-perf/blob/turbo-updates/data-types.md#object-properties)?

That could be compromise of keeping those links out of your eye but providing an idea of what awaits you when you follow it.